### PR TITLE
Added Jaeger Propagator to Opentelemetry.Extensions.Propagators

### LINKED ISSUE #1881: Support for Jaeger Propagator
# Feature Request

The [OpenTelemetry specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#propagators-distribution) suggests that a propagator for Jaeger MUST be implemented alongside the B3 (Zipkin) and W3C Trace Context.

I cannot seem to find any traces of such a Propagator being implemented and it would be great if it was :) Some third party applications like Nginx, used in many proxies rely on this format and valuable trace information can be lost in that step.

**Is your feature request related to a problem?**

Some application, typically third party, rely on "older" Opentracing implementation using Jaeger tracing context and this information is not being tied together with the current implementation of the dotnet library.

**Describe the solution you'd like:**

Adding a propagator that implements, understands and propagate the [Jaeger (uber header)](https://www.jaegertracing.io/docs/latest/client-libraries/#propagation-format) format.

If this has been implemented and I cannot find it, I apologies, I just cannot seem to find any information about it.

Thanks
Kristoffer

### PR DESCRIPTION
**Pull request by notcool11**
* added jaeger progagator with tests

* updated changelog

* moved comment to unreleased

Fixes #1881 .

## Changes

Added Jaeger Propagator to the new Opentelemetry.Extensions.Propagators project with tests. Been running this code locally as a standalone propagator in production for about six months.

Couple of notes, I've found that URL encoding in headers is not consistent so sometimes the `uber-trace-id` is delimited with a `:` colon and sometimes encoded to `%3A`. There is also code to handle the shorter 64 traceid vs. the newer 128 bit format. 

Unit tests should cover all these items. Thanks.

For significant contributions please make sure you have completed the following items:

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by reyang** [comment]
More of a question for the maintainers - do we want different propagators to be bundled as a single package or they should be decoupled?

**Comment by notcool11** [comment]
> More of a question for the maintainers - do we want different propagators to be bundled as a single package or they should be decoupled?

Oh please no. :(

The discussion in [1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881) has bounced several times already. I had it separate before [3244](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3244) was merged in and created the new extensions package.

**Comment by cijothomas** [comment]
> More of a question for the maintainers - do we want different propagators to be bundled as a single package or they should be decoupled?

The Non W3C ones, zipkin/jaeger can be a single package. I think @alanwest suggested it that way earlier, and I agree.
The W3C is part of the main package itself.

File: src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md [hunk 1]
```diff
@@ -4,3 +4,5 @@
 
 * Initial release. This has been ported as is from
 [OpenTelemetry.Api](../OpenTelemetry.Api/README.md) package.
+
+* added jaegerpropagator for issue [1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881)
```
Comment (alanwest) [inline src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md:8]: nit

```suggestion
* Added JaegerPropagator ([1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881))
```

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    {
+        if (context.ActivityContext.IsValid())
+        {
+            // If a valid context has already been extracted, perform a noop.
+            return context;
+        }
+
+        if (carrier == null)
+        {
+            OpenTelemetryApiEventSource.Log.FailedToExtractActivityContext(nameof(JaegerPropagator), "null carrier");
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs:55]: we want to start having own eventsource for this new package.

Comment (notcool11) [reply src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs:55]: @cijothomas , can we punt on this?? i'm leaving the company that this PR is associated to this week.

**Comment by notcool11** [comment]
@reyang , @cijothomas , @alanwest 
this was a pet project for me at my current company, but my last day here will be 06/10. i'll not have access to the fork after i leave. i also don't have confidence that anyone else at the company will care enough to try to babysit this PR to make any changes.

File: src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs:55]: To ship the stable version of this, I fixed this to use own eventsource. You need to pick the latest from main to see the changes reflected.

**Comment by pellared** [comment]
@notcool11, are you able to fix the build failures?

**Comment by notcool11** [comment]
> > 
> 
> @pellared , yep. was failing after pulling latest, had to update the event source per @cijothomas . building in my fork now, should be done shortly

Green in my fork now.

**Comment by pellared** [comment]
There are conflicts with `main`

**Comment by notcool11** [comment]
@pellared , ugh, that's lame. i didn't even change that file. stupid refresh from main now shows 129 files changed when really i just have 5. what a pita.

**Comment by pellared** [comment]
```
git remote add upstream https://github.com/open-telemetry/opentelemetry-dotnet.git
git fetch -p upstream
git merge upstream/main
```

**Comment by notcool11** [comment]
only files that matter in this PR:
src/OpenTelemetry.Extensions.Propagators/.publicApi/net462/PublicAPI.Unshipped.txt
src/OpenTelemetry.Extensions.Propagators/.publicApi/netstandard2.0/PublicAPI.Unshipped.txt
src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md
src/OpenTelemetry.Extensions.Propagators/JaegerPropagator.cs 
test/OpenTelemetry.Extensions.Propagators.Tests/JaegerPropagatorTest.cs

do you need me to pull from main again or can you just merge and ignore changes in this PR except the above files?? my guess is this fork will continue to drift and will need to be rebased.

**Comment by notcool11** [comment]
> ```
> git remote add upstream https://github.com/open-telemetry/opentelemetry-dotnet.git
> git fetch -p upstream
> git merge upstream/main # or git rebase upstream/master - but this would require force push
> ```

okay, i'll try that

File: src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md [hunk 3]
```diff
@@ -15,3 +15,5 @@ Released 2022-June-1
 
 * Initial release. This has been ported as is from
 [OpenTelemetry.Api](../OpenTelemetry.Api/README.md) package.
+
+* Added JaegerPropagator ([1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881))
```
[the remark this answers is not in the document]
Comment (notcool11) [reply src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md]: good catch. thanks.

File: src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+* Added JaegerPropagator ([1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881))
```
Comment (pellared) [inline src/OpenTelemetry.Extensions.Propagators/CHANGELOG.md:5]: nit

```suggestion
* Add `JaegerPropagator`.
  ([1881](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1881))
```

**Review by pellared — APPROVED** [review]

**Pull request merged**